File: apps/minecraft/velocity/src/main/java/crabcraft/net/crabUtilities/velocity/staffchat/StaffChatManager.java
package crabcraft.net.crabUtilities.velocity.staffchat;

import com.velocitypowered.api.proxy.Player;
import crabcraft.net.crabUtilities.velocity.CrabUtilitiesVelocity;
import crabcraft.net.crabUtilities.velocity.DiscordWebhook;
import crabcraft.net.crabUtilities.velocity.NicknameComponentParser;
import net.kyori.adventure.text.Component;
import net.kyori.adventure.text.minimessage.MiniMessage;
import net.kyori.adventure.text.minimessage.tag.resolver.Placeholder;
import net.kyori.adventure.text.serializer.plain.PlainTextComponentSerializer;

import java.util.Set;
import java.util.UUID;
import java.util.concurrent.ConcurrentHashMap;

public class StaffChatManager {

    static final String PERMISSION = "crabutilities.staffchat";
    private static final MiniMessage MINI_MESSAGE = MiniMessage.miniMessage();

    private final CrabUtilitiesVelocity plugin;
    private final RedisStaffChat redis;
    private final DiscordWebhook discordWebhook;
    private final String discordAvatarUrlTemplate;
    private final Set<UUID> disabledPlayers = ConcurrentHashMap.newKeySet();

    public StaffChatManager(CrabUtilitiesVelocity plugin, RedisStaffChat redis,
                            DiscordWebhook discordWebhook, String discordAvatarUrlTemplate) {
        this.plugin = plugin;
        this.redis = redis;
        this.discordWebhook = discordWebhook;
        this.discordAvatarUrlTemplate = discordAvatarUrlTemplate;
    }

    public boolean isEnabled(UUID uuid) {
        return !disabledPlayers.contains(uuid);
    }

    public boolean toggle(UUID uuid) {
        if (disabledPlayers.remove(uuid)) {
            return true;
        } else {
            disabledPlayers.add(uuid);
            return false;
        }
    }

    public void sendMessage(String senderName, UUID senderUuid, String message) {
        redis.publish(senderName, message);
        sendToDiscord(senderName, senderUuid, message);
    }

    private void sendToDiscord(String senderName, UUID senderUuid, String message) {
        if (discordWebhook == null) return;
        String plainName = NicknameComponentParser.plain(senderName);
        String avatarUrl = null;
        if (senderUuid != null && discordAvatarUrlTemplate != null && !discordAvatarUrlTemplate.isEmpty()) {
            avatarUrl = discordAvatarUrlTemplate.replace("{uuid}", senderUuid.toString());
        }
        discordWebhook.send(message, plainName, avatarUrl);
    }

    public void displayMessage(String senderName, String message) {
        String format = plugin.getConfig().getStaffChatFormat();
        Component senderComponent = NicknameComponentParser.parse(senderName);
        Component component = MINI_MESSAGE.deserialize(format,
                Placeholder.component("sender", senderComponent),
                Placeholder.unparsed("message", message)
        );

        for (Player player : plugin.getServer().getAllPlayers()) {
            if (player.hasPermission(PERMISSION) && isEnabled(player.getUniqueId())) {
                player.sendMessage(component);
            }
        }

        String plainSender = PlainTextComponentSerializer.plainText().serialize(senderComponent);
        plugin.getLogger().info("[StaffChat] {}: {}", plainSender, message);
    }
}

File: apps/minecraft/velocity/src/main/kotlin/crabcraft/net/crabUtilities/velocity/staffchat/StaffChatManager.kt
package crabcraft.net.crabUtilities.velocity.staffchat

import com.velocitypowered.api.proxy.Player
import crabcraft.net.crabUtilities.velocity.CrabUtilitiesVelocity
import crabcraft.net.crabUtilities.velocity.DiscordWebhook
import crabcraft.net.crabUtilities.velocity.NicknameComponentParser
import net.kyori.adventure.text.Component
import net.kyori.adventure.text.minimessage.MiniMessage
import net.kyori.adventure.text.minimessage.tag.resolver.Placeholder
import net.kyori.adventure.text.serializer.plain.PlainTextComponentSerializer
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap

open class StaffChatManager(private val plugin: CrabUtilitiesVelocity, private val redis: RedisStaffChat,
                            private val discordWebhook: DiscordWebhook?, private val discordAvatarUrlTemplate: String?) {
    private val disabledPlayers = ConcurrentHashMap.newKeySet<UUID>()
    open fun hasPermission(player: Player) = player.hasPermission(PERMISSION)
    open fun isEnabled(uuid: UUID) = !disabledPlayers.contains(uuid)
    open fun toggle(uuid: UUID): Boolean {
        if (disabledPlayers.remove(uuid)) return true
        disabledPlayers.add(uuid)
        return false
    }
    open fun sendMessage(senderName: String, senderUuid: UUID?, message: Component) {
        redis.publish(senderName, message)
        sendToDiscord(senderName, senderUuid, message)
    }
    private fun sendToDiscord(senderName: String, senderUuid: UUID?, message: Component) {
        if (discordWebhook == null) return
        val plainName = NicknameComponentParser.plain(senderName)
        val plainMessage = PlainTextComponentSerializer.plainText().serialize(message)
        var avatarUrl: String? = null
        if (senderUuid != null && !discordAvatarUrlTemplate.isNullOrEmpty()) avatarUrl = discordAvatarUrlTemplate.replace("{uuid}", senderUuid.toString())
        discordWebhook.send(plainMessage, plainName, avatarUrl)
    }
    open fun displayMessage(senderName: String, message: Component) {
        val format = plugin.getConfig().getStaffChatFormat()
        val senderComponent = NicknameComponentParser.parse(senderName)
        val component = MINI_MESSAGE.deserialize(format, Placeholder.component("sender", senderComponent), Placeholder.component("message", message))
        for (player in plugin.getServer().allPlayers) if (player.hasPermission(PERMISSION) && isEnabled(player.uniqueId)) plugin.getMessageManager().deliver(player, component)
        val plainSender = PlainTextComponentSerializer.plainText().serialize(senderComponent)
        val plainMessage = PlainTextComponentSerializer.plainText().serialize(message)
        plugin.getLogger().info("[StaffChat] {}: {}", plainSender, plainMessage)
    }
    companion object {
        const val PERMISSION = "crabutilities.staffchat"
        private val MINI_MESSAGE = MiniMessage.miniMessage()
    }
}
